catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -37,41 +37,84 @@
  * directory being monitored, and call the {@link AsyncFileAlterationListener}'s corresponding
  * methods
  *
- * @apiNote The AsyncFileAlterationListener is expected to either call the callback function passed
- *     with {@code null} or pass the callback function into an {@link org.apache.camel.Exchange}.
- * @apiNote If this is loaded from a file after the system has been shut down, the user
- *     <strong>MUST</strong> call {@code clearCache()} in order to clear the currently processing
- *     request.
+ * @see AsyncFileAlterationListener
  */
 public class AsyncFileAlterationObserver implements Serializable {
 
-  private final AsyncFileEntry rootFile;
-  private AsyncFileAlterationListener listener = null;
-  private final Set<AsyncFileEntry> processing = new ConcurrentSkipListSet<>();
+  @Nullable private final AsyncFileEntry rootFile; //  In case GSON returns it.
+  @Nullable private transient AsyncFileAlterationListener listener = null;
+  private final transient Set<AsyncFileEntry> processing = new ConcurrentSkipListSet<>();
   private static final Logger LOGGER = LoggerFactory.getLogger(AsyncFileAlterationObserver.class);
+  private final transient Object listenerLock = new Object();;
 
   public AsyncFileAlterationObserver(File fileToObserve) {
+    if (fileToObserve == null) {
+      throw new IllegalArgumentException();
+    }
     rootFile = new AsyncFileEntry(fileToObserve);
   }
 
-  //  Only for testing
+  //  For GSON serialization;
+  private AsyncFileAlterationObserver() {
+    rootFile = null;
+  }
+
+  public void onLoad() {
+    if (rootFile != null) rootFile.initialize();
+  }
+
+  //  Returns true on errors;
+  private boolean initHelper(AsyncFileEntry entry) {
+    File[] children = listFiles(entry.getFile());
+    if (children == null) {
+      //  Error while initializing... some records may appear twice.
+      LOGGER.debug("Error getting children for [{}]", entry.getName());
+      return true;
+    }
+    boolean errors = false;
+    for (File child : children) {
+      AsyncFileEntry childEntry = new AsyncFileEntry(entry, child);
+      entry.addChild(childEntry);
+      errors = errors || initHelper(childEntry);
+    }
+    return errors;
+  }
+
+  /**
+   * Initializes the object state of the Observer.
+   *
+   * @return true on success, false if an IO error occurs
+   */
+  public boolean initialize() {
+    return !initHelper(rootFile);
+  }
+
+  public void destroy() {
+    rootFile.destroy();
```

Comment:
❓ Do we need to perform null checks since rootFile is nullable?

Reply:
It's really only nullable if it's not in the json. I'm going to throw an `IllegalStateException` in the `onLoad` method (or whatever that might turn into that's called after initializing) to prevent that from being null during proper use. I feel like this is better than putting null checks everywhere but I very well could be wrong